File: app/assets/js/contracts/crowdfund.ts
export const getContract = async (eos: any) => {
  return await eos.contract('crowdfund')
}

export type CreateCampaignInput = {
  categoryDetails: {
    category: string
  },
  projectDetails: {
    title: string,
    description: string,
    image: string,
  },
  campaignDetails: {
    fundingGoal: number,
    campaignStart: number,
    campaignEnd: number
  },
  rewardDetails: {
    title: string,
    pledgeAmount: number,
    description: string,
    estimatedDelivery: number
  }
}

type RemoveCrowdfundInput = {
  id: number
}

export const createCampaign = (contract: any) => async (input: CreateCampaignInput) => {
  await contract.ccampaign({ input: {
    creator: "crowdfund",
    category: input.categoryDetails.category,
    project_title: input.projectDetails.title,
    description: input.projectDetails.description,
    image: input.projectDetails.image,
    funding_goal: input.campaignDetails.fundingGoal,
    campaign_start: input.campaignDetails.campaignStart,
    campaign_end: input.campaignDetails.campaignEnd,
    title: input.rewardDetails.title,
    pledge_amount: input.rewardDetails.pledgeAmount,
    reward_description: input.rewardDetails.description,
    estimated_delivery: input.rewardDetails.estimatedDelivery,
  } }, { authorization: "crowdfund" })
}

export const removeCrowdfund = async (contract: any, input: RemoveCrowdfundInput) => {
  await contract.rcampaign(input.id - 1), { authorization: "crowdfund" }
}

File: contract/crowdfund/crowdfund.cpp
#include "crowdfund.hpp"

#include <eosiolib/eosio.hpp>

using namespace eosio;
using std::string;

namespace coinstarter {

  crowdfund::crowdfund( account_name s )
  :contract(s),
  campaigns(_self, _self)
  {};

  void crowdfund::ccampaign(const campaigninput input) {
    campaigns.emplace(_self, [&](campaign& newCampaign) {
      newCampaign.id = campaigns.available_primary_key();
      newCampaign.creator = input.creator;
      newCampaign.category = input.category;
      newCampaign.project_title = input.project_title;
      newCampaign.description = input.description;
      newCampaign.image = input.image;
      newCampaign.funding_goal = input.funding_goal;
      newCampaign.campaign_start = input.campaign_start;
      newCampaign.campaign_end = input.campaign_end;
      newCampaign.title = input.title;
      newCampaign.pledge_amount = input.pledge_amount;
      newCampaign.reward_description = input.reward_description;
      newCampaign.estimated_delivery = input.estimated_delivery;
    });
  }

  void crowdfund::rcampaign(const uint64_t campaignId) {
    auto itr = campaigns.find(campaignId);
    if (itr != campaigns.end())
    {
      campaigns.erase(itr);
    }
  }

  EOSIO_ABI( crowdfund, (ccampaign) (rcampaign) )
}
